### PR TITLE
Github Action for Automated Release

### PR DESCRIPTION
**Pull request by AkMo3**
Meta issue corresponding to [Akmo3/Cytoscape.js](https://github.com/AkMo3/cytoscape.js/) fork https://github.com/AkMo3/cytoscape.js/issues/2

Steps required before merging:
1. Create an environment named `prod` for the secrets to be stored. 
2. Creation of secrets. The secrets are npmjs release secret, github secret for repositories [Cytoscape.js](https://github.com/cytoscape/cytoscape.js) and  [Cytoscape.js-blog](https://github.com/cytoscape/cytoscape.js)
3. Add these secrets to the environment.

### CODE REVIEW

**Comment by maxkfranz** [comment]
```
DIRNAME=$(realpath `dirname "$0"`)
ROOTDIR=$(realpath "$DIRNAME/../../")
```

File: package.json [hunk 1]
```diff
@@ -105,5 +104,8 @@
   "dependencies": {
     "heap": "^0.2.6",
     "lodash": "^4.17.21"
+  },
+  "directories": {
+    "test": "test"
```
Comment (AkMo3) [inline package.json:109]: Used to specify where all test code is present. (Automatically added during the release process by npm).

File: .github/workflows/feature-release-test.yml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+          jq --version
+        shell: bash
+      - name: Checkout Master Branch
+        uses: actions/checkout@v3
+        with:
+          ref: master
+          fetch-depth: 0
+      - name: Merge unstable to master branch
+        run: |
+          curl -L -o merge_unstable_to_master.sh https://raw.githubusercontent.com/AkMo3/cytoscape.js/unstable/.github/workflows/scripts/merge_unstable_to_master.sh
```
[the remark this answers is not in the document]
Comment (AkMo3) [reply .github/workflows/feature-release-test.yml]: Resolved

File: .github/workflows/patch-release-test.yml [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+          ref: ${{ env.BRANCH }}
+      - uses: actions/setup-node@v3
+        with:
+          node-version: 18
+          cache: 'npm'
+          registry-url: 'https://registry.npmjs.org'
+      - name: Get new version string
+        id: get_new_patch_version
+        run: |
+          cp ./.github/workflows/scripts/new-patch-version.sh new-patch-version.sh
```
[the remark this answers is not in the document]
Comment (AkMo3) [reply .github/workflows/patch-release-test.yml]: Resolved

File: .github/workflows/patch-release-test.yml [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        uses: actions/upload-artifact@v3
+        if: ${{ failure() && steps.release.conclusion == 'failure' }}
+        with:
+          name: npm-release--failure-report
+          path: /home/runner/.npm/_logs/
+      - name: Publish Package To npmjs
+        run:  npm publish
+        env:
+          NODE_AUTH_TOKEN: ${{ secrets.CYTOSCAPE_PUBLISH_TOKEN }}
+      - name: Publish Package to GitHub Releases
```
[the remark this answers is not in the document]
Comment (AkMo3) [reply .github/workflows/patch-release-test.yml:103]: Publishing should be done for all releases. This test condition should be applied while pushing documentation for 'back-port' releases.

File: .gitignore [hunk 5]
```diff
@@ -11,3 +11,10 @@ documentation/slimerjs-*/**
 documentation/api/**
 documentation/download/*.zip
 benchmark/suite/cytoscape.js
+
+# CI/CD Files
+New-Feature-Version.js
```
[the remark this answers is not in the document]
Comment (AkMo3) [reply .gitignore]: Done

File: .gitignore [hunk 6]
```diff
@@ -11,3 +11,10 @@ documentation/slimerjs-*/**
...
+merge_unstable_to_master.sh
+new-feature-version.sh
+pre_release_test.sh
+new-pathc-version.sh
```
[the remark this answers is not in the document]
Comment (AkMo3) [reply .gitignore]: Done

**Review by AkMo3 — COMMENTED** [review]
@maxkfranz Review completed from my side, all changes have been done

**Comment by maxkfranz** [comment]
Great.  Thanks, Akash.  I’ll review it tomorrow.

**Comment by maxkfranz** [comment]
Notes:

- The names of the secrets aren't clear.  CYTOSCAPE_PUBLISH_TOKEN and CYTOSCAPE_JS_TOKEN are confusing.  Something like this would be better:
  - NPM_TOKEN : the token for npm
  - GITHUB_TOKEN : the token for GitHub
- The documentation should link to resources, or include steps within the documentation, so that the process of creating the above tokens is explicit.
- The actions are titled things like 'Patch Release Test'.  They should all have final names like 'Patch Release'.
- When I press the 'Run Workflow' button, I see two fields:
  - (1) Use workflow from [unstable]
  - (2) Define branch name [unstable]
  - If they're the same thing, you shouldn't have to specify it twice.  That introduces opportunity for user error.
- I tried running the patch release workflow on master in this fork of your fork: 
  - Fork: https://github.com/maxkfranz/cytoscape.js-auto-release-test
  - Result: https://github.com/maxkfranz/cytoscape.js-auto-release-test/actions/runs/5956616717
  - The workflow was started with parameters master/master. 
  - It looks like there is an error.

**Comment by AkMo3** [comment]
1. 
> I tried running the patch release workflow on master in this fork of your fork:
> Fork: https://github.com/maxkfranz/cytoscape.js-auto-release-test
>    Result: https://github.com/maxkfranz/cytoscape.js-auto-release-test/actions/runs/5956616717
>    The workflow was started with parameters master/master.
>    It looks like there is an error.

The error occurred because the GitHub action couldn't access the read/write on the repository. We need to give that explicitly in the setting tab. Ref: [managing-github-actions-permissions-for-your-repository](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#managing-github-actions-permissions-for-your-repository)

2. 
> The workflow was started with parameters master/master.

The default branch of the repository is `master`. That's why the `master` is shown.

3. 
> When I press the 'Run Workflow' button, I see two fields:
    (1) Use workflow from [unstable]
    (2) Define branch name [unstable]
    If they're the same thing, you shouldn't have to specify it twice. That introduces opportunity for user error.
    
 They are different. One field is a default field that GitHub provides to select from the branch you would like to run your action (default to your default branch), and the `Define branch name` is to select on which branch you want to run the action. We created this field to select the branch and run the latest workflows in `unstable`.

**Comment by AkMo3** [comment]
> - The names of the secrets aren't clear. CYTOSCAPE_PUBLISH_TOKEN and CYTOSCAPE_JS_TOKEN are confusing. Something like this would be better:
    NPM_TOKEN : the token for npm
    GITHUB_TOKEN : the token for GitHub
 - The documentation should link to resources, or include steps within the documentation, so that the process of creating the above tokens is explicit.
 - The actions are titled things like 'Patch Release Test'. They should all have final names like 'Patch Release'.

Changes done

**Comment by AkMo3** [comment]
If you are running on a fork, you need to make the following changes:
1. Give access to GitHub Actions to read and write on the repository
2. Create a corresponding npm package, and change the `package.json` and `package-lock.json`
3. Change the repository and organization name in GitHub action
4. Create access tokens accordingly

**Comment by maxkfranz** [comment]
Notes:

- If there are any configurations that are necessary that aren't explicitly documented in the readme/md files, then they need to be documented.
 - R/W
 - Releases
 - Secrets
 - NPM automation token 
 - Github token: create the token under the cy org
 - Github token exp. : 
   - set as far in the future as possible (1 years)
   - Permissions
   - Actions: r/w
   - Contents (releases): r/w
 - Tokens: NPM token, main repo token (e.g. release), blog issue token

**Comment by maxkfranz** [comment]
Successfully (mostly) ran a patch release in this fork, to reproduce the workflow from scratch: https://github.com/maxkfranz/cytoscape.js-auto-release-test

**Comment by maxkfranz** [comment]
TODO: Once the feedback has been addressed, let's run the the other actions in the fork repo:

- [x] Patch release
- [ ] Back-port release
- [ ] Feature release

List out feedback, if any, from running these two workflows below.

**Comment by maxkfranz** [comment]
@AkMo3, also note that the URL you submit to GSOC as your work item should meet the following requirements: https://developers.google.com/open-source/gsoc/help/work-product

You could use this PR as the URL if you edit the text accordingly.

**Comment by AkMo3** [comment]
Will complete all the remaining things before tomorrow meet. As I said in the prev meet, I have full time course going on along with me remote intern so I very limited time right now.

**Comment by maxkfranz** [comment]
OK, talk with you tomorrowOn Aug 30, 2023, at 14:48, Akash Mondal ***@***.***> wrote:﻿
Will complete all the remaining things before tomorrow meet. As I said in the prev meet, I have full time course going on along with me remote intern so I very limited time right now.

**Comment by AkMo3** [comment]
All required changes done

**Comment by maxkfranz** [comment]
Notes:

- [ ] Update version text box label so that it's clear that it's optional.  The version is set automatically to bump the second digit.  However, you can override when needed (e.g. major version, like v4). 
- [ ] https://github.com/maxkfranz/cytoscape.js-auto-release-test/blob/unstable/.github/workflows/feature-release.yml#L11C3-L11C23
- [ ] ...

File: .github/workflows/create-blog-issue.yml [hunk 7]
```diff
@@ -0,0 +1,28 @@
+name: Create Issue in Blog Repo
+on:
+  workflow_dispatch
+
+env:
+  VERSION: 3.26.0
```
Comment (maxkfranz) [inline .github/workflows/create-blog-issue.yml:6]: Hardcoded?

Comment (AkMo3) [reply .github/workflows/create-blog-issue.yml]: Removed

**Comment by AkMo3** [comment]
@maxkfranz I tested the backport release, twice. And it works as expected. 

Now coming to the backport version that you saw as unexpected because unsynced branches. 
Brief explanation: The when you made the release the master was not update with the unstable.

In detail explanation:
When you made the release the `unstable` branch was at commit https://github.com/maxkfranz/cytoscape.js-auto-release-test/commit/21ba489f4a89ea70d4f81abf112ad3e49d4227a4
image

The version present inside package.json is 3.42.0-unstable
image

When you look at action run logs the `master` head is at https://github.com/maxkfranz/cytoscape.js-auto-release-test/commit/0abd6d090ec2d360939683ef84acb8b6a54b42ab with the parent commit being https://github.com/maxkfranz/cytoscape.js-auto-release-test/commit/1043986753711aef223cf2471963f84ad945866f
image

The version present in `master` is 3.40.0
image

So when the backport branch was created, it had version 3.40.0
image


This was actually verified the last feature release where 3.49.0 was on the master initally, and then the backport branch 3.49.x was created where the version was 3.49.0 and master was bumped to 3.50.9

**Comment by maxkfranz** [comment]
Great. It looks like you have a handle on creating a clean environment for properly testing this. Let's review tomorrow

**Pull request merged**